### PR TITLE
Look into optimizing out the hashtable lookups from nsContainerFrame::DestroyFrom and nsContainerFrame::SafelyDestroyFrameListProp

### LINKED ISSUE #1154: Look into optimizing out the hashtable lookups from nsContainerFrame::DestroyFrom and nsContainerFrame::SafelyDestroyFrameListProp
https://bugzilla.mozilla.org/show_bug.cgi?id=1367206

I think all these child list properties are rare so iterating over
the properties once upfront instead of four separate lookups
should be a win. Now that  #1151 is merged this can be done next.

This further improves results of:

http://sinz.org/Maze/

### PR DESCRIPTION
**Pull request by win7-7**
This further improves performance after merging  #1151 to UXP.

Resolves #1154.

### CODE REVIEW

File: layout/base/FrameProperties.h [hunk 1]
```diff
@@ -241,6 +246,26 @@ class FrameProperties
...
   }
+  
+   /**
+   * Call @aFunction for each property or until @aFunction returns false.
+   */
+  template<class F>
+  void ForEach(F aFunction) const
+  {
+#ifdef DEBUG
+   size_t len = mProperties.Length();
```
Comment (wolfbeast) [inline layout/base/FrameProperties.h]: Wrong indent.
Also, only defining this in debug but using it in an assert outside of debug is bad, see line 261.

File: layout/generic/nsContainerFrame.cpp [hunk 2]
```diff
@@ -219,24 +219,49 @@ nsContainerFrame::DestroyFrom(nsIFrame* aDestructRoot)
...
-  MOZ_ASSERT(IsFrameOfType(nsIFrame::eCanContainOverflowContainers) ||
-             !(GetProperty(nsContainerFrame::OverflowContainersProperty()) ||
-               GetProperty(nsContainerFrame::ExcessOverflowContainersProperty())),
-             "this type of frame should't have overflow containers");
-  SafelyDestroyFrameListProp(aDestructRoot, shell,
-                             OverflowContainersProperty());
-  SafelyDestroyFrameListProp(aDestructRoot, shell,
-                             ExcessOverflowContainersProperty());
+  MOZ_ASSERT(IsFrameOfType(eCanContainOverflowContainers) ||
+                           !(hasOC || hasEOC),
```
Comment (wolfbeast) [inline layout/generic/nsContainerFrame.cpp]: This is a very confusing logical operation; please rewrite to something more sensible.

**Comment by wolfbeast** [comment]
Also, please make sure you use proper descriptive commit messages and pull request titles/descriptions -- as-is it's not acceptable because it will be impossible to re-find messages like this in the commit history on keywords.

**Comment by win7-7** [comment]
Is this correct way to do first change requested/ defining?

   template<class F>
   void ForEach(F aFunction) const
   {
#ifdef DEBUG
      size_t len = mProperties.Length();
#endif
       for (const auto& prop : mProperties) {
          bool shouldContinue = aFunction(prop.mProperty, prop.mValue);
          size_t len = mProperties.Length();
          MOZ_ASSERT(len == mProperties.Length(),

Sorry I do not have proper knowledge how to fix these requested changes. There is nothing similar from Mozilla that could be used as to fix these parts. Mozilla files of these still use these same way as in this PR.

Could you also tell exactly how second change you requested should be done?

**Comment by wolfbeast** [comment]
Look at your code
```
#ifdef DEBUG
size_t len = mProperties.Length();
#endif
```
len is only defined if you are in debug mode
A few lines below:
```
MOZ_ASSERT(len == mProperties.Length()
```
Using len in a check that is *not* in debug mode, so `len` will be undefined.
Just adding a definition right above it is pointless since the assertion will then always be true. The point is to check inside the loop whether `mProperties.Length()` has changed from the previous `len`.
Just wrap the assert in `#ifdef DEBUG` PP statements.

The second one is fine as-is, I guess. Not happy with the logic sequence but it works.

**Comment by win7-7** [comment]
I will close this pull request to do new pull request instead.

**Pull request closed without merging**